discovery_engine_core/core/src/ranker.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use xayn_ai::{ranker::Embedding, DocumentId, UserFeedback};
+
+use crate::{
+    document::{Document, Id, TimeSpent, UserReacted, UserReaction},
+    engine::GenericError,
+};
+
+/// Provides a method for ranking a slice of [`Document`] items.
+pub trait Ranker {
+    /// Performs the ranking of [`Document`] items.
+    fn rank(&mut self, items: &mut [Document]) -> Result<(), GenericError>;
+
+    /// Learn from the time a user spent on a document.
+    fn time_logged(&mut self, time_logged: &TimeSpent) -> Result<(), GenericError>;
+
+    /// Learn from a user's interaction.
+    fn user_reacted(&mut self, reaction: &UserReacted) -> Result<(), GenericError>;
```

Comment:
should we adapt the naming of `xayn_ai`'s `Ranker` for the trait as well?
```suggestion
    /// Learn from the time a user spent on a document.
    fn log_document_view_time(&mut self, time_spent: &TimeSpent) -> Result<(), GenericError>;

    /// Learn from a user's interaction.
    fn log_user_reaction(&mut self, reaction: &UserReacted) -> Result<(), GenericError>;
```